### PR TITLE
Feature/new components browser

### LINKED ISSUE #528: Building a Component Browser
* [x] Is a *Parts Bin* like way to quickly browse all available components in the project and other projects (including the core packages).
- The current component browser can be seen as a first sketch
- The components browser is basically prompt that can be *popped out* into a window to be freely moved around. This can be convenient in case more frequent access is required.
  - [x] this warrants the implementation of a "popping window", i.e., a window which can be closed with `esc` when focused (*Now fixed by just displaying the components browser as a popup*)
* [x] Allows to instantiate components with behavior enabled or disabled. This is controlled via a global flag on the bottom of the components browser.
* [x] Components can be instantiated via drag and drop.
* [x] Components can be applied to override certain state based components by dragging the components on UI elements in the side bar or the responsive design halo. (#655)
* [x] Components can be organized by name (alphabetically) or grouped by component files. Also there is always the option for the user to query for a particular component via search.
* [x] Components can also be opened in order to edit the component definition via direct manipulation. This is carried out by clicking the "edit" button on the bottom the the component browser, when a particular component is selected.

#### Importing of Components from other Projects

- Components residing in other (i.e. 3rd party) projects can *not be edited*. However they can be reused in the currently open project to serve as building blocks for new components.
  - everything that is available in the local projects folder can be imported
  - when we use a project that depends on a project we do not have locally, we clone it from github 
  - using a component from another project warrants importing the projects `index.css`

### LINKED ISSUE #534: Component Declaration and Internal Management
- Components are declared via Halo interface.
  - [x] At first the component name declared the file name that is created and the component is defined in.
  - [x] The halo interface ensures that there are not name conflicts within the project.
  - [x] Components without parent components reside in their own file
- All derived components are defined via copy halo interface.
  - [x] Once they are declared to be components, they are placed in the same file as their parent.

### PR DESCRIPTION
**Pull request by merryman**
Closes #528.
Closes #534.

- [x] light grey background
- [x] unify size of items
- [x] list only loaded projects and its dependecies by default
- [x] allow listing of other installed projects via a button -> need to interact with Project API, since we need to ensure existence of deps
- [x] when we add an import to a project, use project API to add dep (reverse is true for removal)
- [x] Renaming top level components still broken, when they dont have a master
- [x] declaring top level components derived from a core library declares them withing core library instead, which is bonkers.
- [x] Adding a composite with name conflicts, does not traverse the introduced submorph hierarchy and renames the introduced morphs in order to avoid name conflicts

### CODE REVIEW

**Comment by linusha** [comment]
1. The preview for the component does not generate
https://github.com/LivelyKernel/lively.next/assets/14252419/dbc18ed9-b627-4f30-8ee6-6d6486079215

2. The component list sorted by name is when not actively searched. This is fine I guess, but I would appreciate some kind of placeholder text like "Begin searching to display components..."

https://github.com/LivelyKernel/lively.next/assets/14252419/82e46c8a-12f4-46e5-9ed5-73ec43fb0412

3. Dropping a new morph into a components, its position does not get reconciled in the "same reconciliation session".

https://github.com/LivelyKernel/lively.next/assets/14252419/7a63085c-51bd-4560-b762-9664789bafe1

4. One cannot edit components from dependency projects via the components browser. The "Open in System Browser" button in this case should be disabled I think (light grey, forbidden cursor, something like that).
Peek 2023-07-03 12-32

5. The "Import" Button should instead read "Open"

File: lively.ide/lively-whitepaper.md
Comment (linusha) [inline lively.ide/lively-whitepaper.md:1]: Not a fan of having this in the repository. We have (an older?) version in the wiki, lets merge them there, ok?

File: lively.ide/studio/component-browser.cp.js [hunk 1]
```diff
@@ -32,10 +41,20 @@ class MasterComponentTreeData extends TreeData {
...
+      return;
+    }
     node.isCollapsed = bool;
     node.isDirty = true;
     if (!bool) {
       if (node.type === 'package') {
+        if (!PackageRegistry.ofSystem(System).lookup(node.name)) {
+          // if this is a project, load the project
+          await importPackage(node.url);
+        }
```
Comment (linusha) [inline lively.ide/studio/component-browser.cp.js:57]: This confuses me...let's talk about it tomorrow.

File: lively.ide/studio/component-browser.cp.js [hunk 2]
```diff
@@ -277,11 +350,18 @@ class MasterComponentTreeData extends TreeData {
...
+    const files = arr.compact(await Promise.all(resources.map(async res => {
       let type;
-      if (res.isDirectory()) type = 'directory';
-      else type = 'cp.js';
+      if (res.isDirectory()) {
+        type = 'directory';
+        if ((await this.listComponentFilesInDir(res.url)).length === 0) return;
+      } else {
+        type = 'cp.js';
+        if ((await res.read()).match(/['"]skip listing['"];/)) return;
```
Comment (linusha) [inline lively.ide/studio/component-browser.cp.js:363]: We should remember documenting this somewhere.
Pretty nice btw!

File: lively.ide/studio/component-browser.cp.js [hunk 3]
```diff
@@ -322,16 +401,35 @@ export class ExportedComponent extends Morph {
         }
       },
       isSelected: {},
+      isInLocalProject: {
```
Comment (linusha) [inline lively.ide/studio/component-browser.cp.js:404]:
```suggestion
      isInOpenedProject: {
```
What do you think?

File: lively.ide/studio/component-browser.cp.js [hunk 4]
```diff
@@ -1046,36 +1358,59 @@ const ProjectSection = component({
     textAndAttributes: ['Project Name', {
     }, '  ', {
     }, '', {
-      fontFamily: '"Font Awesome 6 Free", "Font Awesome 6 Brands"',
+      fontFamily: '"Font Awesome 5 Free", "Font Awesome 5 Brands"',
```
Comment (linusha) [inline lively.ide/studio/component-browser.cp.js:1361]: ?

File: lively.ide/studio/component-browser.cp.js [hunk 5]
```diff
@@ -1046,36 +1358,59 @@ const ProjectSection = component({
...
       fontWeight: '900',
       nativeCursor: 'pointer',
       paddingTop: '3px',
+      textStyleClasses: ['fas']
```
Comment (linusha) [inline lively.ide/studio/component-browser.cp.js:1365]: This does not work anymore (or at least will not).

File: lively.ide/studio/component-browser.cp.js [hunk 6]
```diff
@@ -1088,80 +1423,405 @@ const ComponentBrowser = component(LightPrompt, {
...
+      fontColor: Color.rgba(0, 0, 0, 0.5),
+      cursorWidth: 1.5,
+      fixedWidth: true,
+      padding: rect(1, 1, 0, 0),
+      scale: 1.32,
+      textAndAttributes: ['', {
+        fontFamily: '"Font Awesome 5 Free", "Font Awesome 5 Brands"',
+        fontWeight: '900',
+        lineHeight: 1,
+        textStyleClasses: ['fas']
```
Comment (linusha) [inline lively.ide/studio/component-browser.cp.js:1458]: See above

File: lively.ide/studio/component-browser.cp.js [hunk 7]
```diff
@@ -1088,80 +1423,405 @@ const ComponentBrowser = component(LightPrompt, {
...
+      visible: false,
+      fontColor: Color.rgba(0, 0, 0, 0.5),
+      fontSize: 25,
+      lineHeight: 2,
+      padding: rect(1, 1, 9, 0),
+      textAndAttributes: ['', {
+        fontFamily: '"Font Awesome 5 Free", "Font Awesome 5 Brands"',
+        fontWeight: '900',
+        lineHeight: 1,
+        textStyleClasses: ['fas']
```
Comment (linusha) [inline lively.ide/studio/component-browser.cp.js:1489]: See above

File: lively.ide/studio/component-browser.cp.js [hunk 8]
```diff
@@ -1088,80 +1423,405 @@ const ComponentBrowser = component(LightPrompt, {
...
+      extent: pt(80, 23.8),
       submorphs: [{
         name: 'label',
-        textAndAttributes: ['IMPORT', null]
+        textAndAttributes: ['', {
+          fontColor: Color.rgb(74, 174, 79),
+          fontFamily: '"Font Awesome 5 Free", "Font Awesome 5 Brands"',
+          fontWeight: '900',
+          lineHeight: 1,
+          textStyleClasses: ['fas']
```
Comment (linusha) [inline lively.ide/studio/component-browser.cp.js:1596]: Once more

File: lively.ide/studio/component-browser.cp.js [hunk 9]
```diff
@@ -1088,80 +1423,405 @@ const ComponentBrowser = component(LightPrompt, {
...
+          master: ButtonDarkDefault,
+          submorphs: [{
+            name: 'label',
+            fontColor: Color.rgb(255, 255, 255),
+            textAndAttributes: ['', {
+              fontColor: Color.rgbHex('B2EBF2'),
+              fontFamily: '"Font Awesome 5 Free", "Font Awesome 5 Brands"',
+              fontWeight: '900',
+              lineHeight: 1,
+              textStyleClasses: ['fas']
```
Comment (linusha) [inline lively.ide/studio/component-browser.cp.js:1725]: baby one more time (these are more mementos for me, feel free to ignore them)

File: lively.ide/studio/component-browser.cp.js [hunk 10]
```diff
@@ -1088,80 +1423,405 @@ const ComponentBrowser = component(LightPrompt, {
...
+          name: 'import button',
+          master: ButtonDarkDefault,
+          submorphs: [{
+            name: 'label',
+            textAndAttributes: ['', {
+              fontColor: Color.rgbHex('B2EBF2'),
+              fontFamily: '"Font Awesome 5 Free", "Font Awesome 5 Brands"',
+              fontWeight: '900',
+              lineHeight: 1,
+              textStyleClasses: ['fas']
```
Comment (linusha) [inline lively.ide/studio/component-browser.cp.js:1740]: :eyes:

File: lively.morphic/rendering/morphic-default.js [hunk 11]
```diff
@@ -264,7 +264,7 @@ textarea.lively-text-input.debug {
 }
 
 .Hand {
-  z-index: 1;
+  z-index: 5;
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.morphic/rendering/morphic-default.js:267]: Done.

File: lively.components/buttons.js [hunk 12]
```diff
@@ -108,6 +108,7 @@ export class ButtonModel extends ViewModel {
 
   trigger () {
     try {
+      signal(this.view, 'fire');
       signal(this, 'fire');
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.components/buttons.js:112]: Pls comment

File: lively.ide/studio/component-browser.cp.js (hunk 1 above)
Comment (linusha) [reply lively.ide/studio/component-browser.cp.js:57]: Should go

File: lively.components/buttons.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (merryman) [reply lively.components/buttons.js:112]: We should and the right way is to signal the view.
But doing it here will require me to go through _a lot_ of files that are not related to this already large PR.
So I would like to do this in another branch in the near future.

File: lively.ide/code-search.cp.js [hunk 13]
```diff
@@ -10,7 +10,11 @@ import { Label } from 'lively.morphic/text/label.js';
 const ModeButtonInactive = component(ButtonDefault, {
   extent: pt(27, 27),
   borderStyle: 'none',
-  fill: Color.transparent
+  fill: Color.transparent,
+  submorphs: [{
+    name: 'label',
+    textAndAttributes: ['B', null]
```
[the remark this answers is not in the document]
Comment (merryman) [reply lively.ide/code-search.cp.js:16]: I will remove this one.

**Comment by linusha** [comment]
- Image does not get added to component via reconciliation
- Revert button resets code but does leave component instance in world with pink halo
- component preview does still not generate

File: lively.components/buttons.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (linusha) [reply lively.components/buttons.js:112]: Yes, we talked about this. I agree, but let's please leave a comment directly in the code...

**Review by linusha — APPROVED** [review]

**Pull request merged**